### PR TITLE
feat: Implement OptimizelyConfig API

### PR DESCRIPTION
**Pull request by pawels-optimizely**
## Summary 

* Implemented OptimizelyConfig API for go sdk
* Performed real tests on datafile taken from swift tests.

### CODE REVIEW

File: pkg/client/optimizely_config.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	for _, feature := range features {
+
+		var optlyFeatureVariablesMap = map[string]OptimizelyVariable{}
+		for _, featureVarible := range feature.VariableMap {
+			optlyVariable := OptimizelyVariable{Key: featureVarible.Key, ID: featureVarible.ID,
+				Type: string(featureVarible.Type), Value: featureVarible.DefaultValue}
+			optlyFeatureVariablesMap[featureVarible.Key] = optlyVariable
+		}
+
+		optlyExperimentMap := getExperimentMap(features, feature.FeatureExperiments, variableByIDMap)
```
Comment (jaeopt) [inline pkg/client/optimizely_config.go]: We do not need to rebuild map here. You can consider re-using experimentsMap built before and pick for each feature's experimentIds.

File: pkg/client/optimizely_config_test.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+	optimizelyFactory := &OptimizelyFactory{}
+	s.optimizelyClient, err = optimizelyFactory.Client(WithConfigManager(projectConfigManager))
+	if err != nil {
+		s.Fail("unable to initialize optimizely client")
+	}
+
+}
+
+func (s *OptimizelyConfigTestSuite) TestOptlyConfig() {
```
Comment (jaeopt) [inline pkg/client/optimizely_config_test.go]: add test case for returning null when config not ready

File: pkg/client/optimizely_config_test.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+		s.Fail("error opening file:" + dataFileName)
+	}
+
+	projectConfigManager, e := config.NewStaticProjectConfigManagerFromPayload(datafile)
+	if e != nil {
+		s.Fail("error constructing config manager")
+	}
+
+	outputFileName := "testdata/optimizely_config_expected.json"
+	expectedOutput, er := ioutil.ReadFile(outputFileName)
```
Comment (jaeopt) [inline pkg/client/optimizely_config_test.go]: I understand this file read and map guarantees that all data fields are public, so no need for extra access control validation. If it's intention, add comments to clarify it.

Comment (pawels-optimizely) [reply pkg/client/optimizely_config_test.go]: comments provided

File: pkg/client/optimizely_config_test.go (hunk 2 above)
Comment (pawels-optimizely) [reply pkg/client/optimizely_config_test.go]: test case added

File: pkg/client/optimizely_config.go (hunk 1 above)
Comment (pawels-optimizely) [reply pkg/client/optimizely_config.go]: addressed

File: pkg/config/datafileprojectconfig/optimizely_config.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+ *                                                                          *
+ * Unless required by applicable law or agreed to in writing, software      *
+ * distributed under the License is distributed on an "AS IS" BASIS,        *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ * See the License for the specific language governing permissions and      *
+ * limitations under the License.                                           *
+ ***************************************************************************/
+
+// Package datafileprojectconfig //
+package datafileprojectconfig
```
Comment (mikeng13) [inline pkg/config/datafileprojectconfig/optimizely_config.go]: This is the wrong package to put it in. The `OptimizelyConfig` shouldn't be tightly coupled to the concept of a `datafile`. If it's the package name confusion that we care about I'd rather us just do `pkg/config/optimizely_config.go` and have the linter ignore it heh. I understand that the package name kind of collides with the file and class names but that is more syntactically correct for us

File: pkg/client/client.go [hunk 5]
```diff
@@ -486,6 +486,17 @@ func (o *OptimizelyClient) GetProjectConfig() (projectConfig config.ProjectConfi
 	return projectConfig, nil
 }
 
+// GetOptimizelyConfig returns OptimizelyConfig object
+func (o *OptimizelyClient) GetOptimizelyConfig() (optimizelyConfig *entities.OptimizelyConfig, err error) {
+
+	projectConfig, err := o.ConfigManager.GetConfig()
```
Comment (mikeng13) [inline pkg/client/client.go]: We can simply call `o.GetProjectConfig()` right? Also, let's make that method private since now everyone should be using this new one.

File: pkg/config/datafileprojectconfig/config.go [hunk 6]
```diff
@@ -161,6 +169,16 @@ func (c DatafileProjectConfig) GetGroupByID(groupID string) (entities.Group, err
...
 }
 
+// GetOptimizelyConfig is the main interface to get OptimizelyConfig object
+func (c DatafileProjectConfig) GetOptimizelyConfig() *entities.OptimizelyConfig {
+
+	featuresList := c.GetFeatureList()
+	experimentsList := c.GetExperimentList()
+	revision := c.GetRevision()
+
+	return entities.NewOptimizelyConfig(featuresList, experimentsList, revision)
```
Comment (mikeng13) [inline pkg/config/datafileprojectconfig/config.go]: @jaeopt should we be caching this lazily instead of re-computing every time? This is tied directly to the project config so we don't have to worry about the cached version becoming stale.

File: pkg/config/datafileprojectconfig/config.go [hunk 7]
```diff
@@ -127,6 +127,14 @@ func (c DatafileProjectConfig) GetFeatureList() (featureList []entities.Feature)
 	return featureList
 }
 
+// GetExperimentList returns an array of all the experiments
+func (c DatafileProjectConfig) GetExperimentList() (experimentList []entities.Experiment) {
```
Comment (mikeng13) [inline pkg/config/datafileprojectconfig/config.go]: Can you add a unit test for this method?

File: pkg/entities/optimizely_config_test.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+)
+
+type OptimizelyConfigTestSuite struct {
+	suite.Suite
+	featureList              []Feature
+	experimentList           []Experiment
+	expectedOptimizelyConfig OptimizelyConfig
+}
+
+var featurelist = `[{"ID":"4482920077","Key":"mutex_group_feature","FeatureExperiments":[{"AudienceIds":[],"ID":"12198292373","LayerID":"12235440722","Key":"experiment_4000",
```
Comment (mikeng13) [inline pkg/entities/optimizely_config_test.go]: Why is this necessary? Is this the same contents from the json test files? Can't we just load those and reuse as json objects?

File: pkg/entities/testdata/optimizely_config_expected.json [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+          "id":"2687470097",
+          "key":"s_foo",
+          "type":"string",
+          "value":"foo"
+        }
+      }
+    }
+  },
+  "revision":"9"
+}
```
Comment (mikeng13) [inline pkg/entities/testdata/optimizely_config_expected.json]: nit: newline

File: pkg/config/datafileprojectconfig/test/optimizely_config_datafile.json [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+      ],
+      "id": "12115595439"
+    }
+  ],
+  "attributes": [],
+  "botFiltering": false,
+  "accountId": "8272261422",
+  "events": [],
+  "revision": "9"
+}
```
Comment (mikeng13) [inline pkg/config/datafileprojectconfig/test/optimizely_config_datafile.json]: nit: newline

File: pkg/config/datafileprojectconfig/test/optimizely_config_expected.json [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+          "id":"2687470097",
+          "key":"s_foo",
+          "type":"string",
+          "value":"foo"
+        }
+      }
+    }
+  },
+  "revision":"9"
+}
```
Comment (mikeng13) [inline pkg/config/datafileprojectconfig/test/optimizely_config_expected.json]: nit: newline

File: pkg/entities/optimizely_config_test.go (hunk 8 above)
Comment (pawels-optimizely) [reply pkg/entities/optimizely_config_test.go]: this is not exactly the same contents. These are just feature list and experiment list. This is used just to perform unit tests here (to keep the coverage). I am testing the whole thing again on project config level.

File: pkg/config/datafileprojectconfig/config.go (hunk 6 above)
Comment (pawels-optimizely) [reply pkg/config/datafileprojectconfig/config.go]: my understanding is that we want to rebuild it every time for now since keeping that in memory will be expensive. But @jaeopt  can confirm.

File: pkg/config/datafileprojectconfig/config.go (hunk 7 above)
Comment (pawels-optimizely) [reply pkg/config/datafileprojectconfig/config.go]: yes, I missed that

File: pkg/config/datafileprojectconfig/config.go (hunk 6 above)
Comment (jaeopt) [reply pkg/config/datafileprojectconfig/config.go]: Yes we discussed the issue in a slack thread. We'll let client cache it for now and will visit again later to see the optimization is needed in the SDK level.

File: pkg/config/polling_manager.go [hunk 12]
```diff
@@ -157,6 +158,7 @@ func (cm *PollingProjectConfigManager) SyncConfig(datafile []byte) {
 	}
 	cmLogger.Debug(fmt.Sprintf("New datafile set with revision: %s. Old revision: %s", projectConfig.GetRevision(), previousRevision))
 	cm.projectConfig = projectConfig
+	cm.optimizelyConfig = NewOptimizelyConfig(projectConfig)
```
Comment (mikeng13) [inline pkg/config/polling_manager.go]: We should do this lazily. So if `cm.optimizelyConfig != nil` then we create the optimizelyConfig

File: pkg/config/polling_manager.go [hunk 13]
```diff
@@ -215,6 +217,13 @@ func (cm *PollingProjectConfigManager) GetConfig() (ProjectConfig, error) {
 	return cm.projectConfig, nil
 }
 
+// GetConfig returns the project config
+func (cm *PollingProjectConfigManager) GetOptimizelyConfig() *OptimizelyConfig {
+	cm.configLock.RLock()
+	defer cm.configLock.RUnlock()
+	return cm.optimizelyConfig
```
Comment (mikeng13) [inline pkg/config/polling_manager.go]: Here we lazily create the optimizelyConfig so that we only ever take up the memory if this method is used.

File: pkg/config/optimizely_config.go [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+
+// Package entities //
+package config
+
+import "github.com/optimizely/go-sdk/pkg/entities"
+
+// OptimizelyConfig is a snapshot of the experiments and features in the project config
+type OptimizelyConfig struct {
+	Revision       string
+	ExperimentsMap map[string]OptimizelyExperiment
```
Comment (mikeng13) [inline pkg/config/optimizely_config.go]: Should these entities types be pointers? Does it matter if we return the pointer to the config?

Comment (pawels-optimizely) [reply pkg/config/optimizely_config.go]: maps are already pointers to the underlying array already so it's ok to have it like that

File: pkg/config/polling_manager.go (hunk 12 above)
Comment (pawels-optimizely) [reply pkg/config/polling_manager.go]: I will address that on the optimizely config

File: pkg/config/polling_manager.go (hunk 13 above)
Comment (pawels-optimizely) [reply pkg/config/polling_manager.go]: this always returns either null or populated optimizely config. it always takes up the memory to construct it every time a datafile is updated.

Comment (pawels-optimizely) [reply pkg/config/polling_manager.go]: ok, we have clarified over IM

File: tests/integration/support/client_wrapper.go [hunk 15]
```diff
@@ -120,6 +122,12 @@ func GetInstance(apiOptions models.APIOptions) *ClientWrapper {
 	return clientInstance
 }
 
+// GetProjectConfig gets the right config project
+func (c *ClientWrapper) GetProjectConfig() config.ProjectConfig {
+	return c.projectConfig
```
Comment (mikeng13) [inline tests/integration/support/client_wrapper.go]: Instead of caching this, we can call `c.client.ConfigManager.GetConfig()`

Comment (pawels-optimizely) [reply tests/integration/support/client_wrapper.go]: ok, addressed

File: pkg/config/polling_manager.go [hunk 16]
```diff
@@ -215,6 +219,18 @@ func (cm *PollingProjectConfigManager) GetConfig() (ProjectConfig, error) {
...
 }
 
+// GetOptimizelyConfig returns the optimizely project config
+func (cm *PollingProjectConfigManager) GetOptimizelyConfig() *OptimizelyConfig {
+	cm.configLock.RLock()
+	defer cm.configLock.RUnlock()
+	if cm.optimizelyConfig != nil {
+		return cm.optimizelyConfig
+	}
+	optimizelyConfig := NewOptimizelyConfig(cm.projectConfig)
```
Comment (jaeopt) [inline pkg/config/polling_manager.go]: This is interesting - triggering the first OptConfig only when called by client. 
Tradeoff between memory saving for most clients who do not use OptConfig vs extra delay for 1st build.
It looks good as is, and we can consider alternate setting based on feedbacks.
